File: in_akatsuki/lazyExam.rb
def div23 n
	if n.even? then div23 (n / 2)
	elsif n % 3 == 0 then div23 (n / 3)
	else n end
end

puts div23(666)

def not23(a, b, c)
	if a.even? then b
	elsif n % 3 then c
	else a end
end

def div23_2 n
	not23(n, div23_2(n / 2), div23_2(n / 3))
end

puts div23_2(666)

def pickOdd(a, b)
	if a.even? then b else a end
end

def div2(n)
	pickOdd(n, div2(n))
end

File: in_akatsuki/lazyExam.hs
div2 :: Int -> Int
div2 n	| even n = div2 (n `div` 2)
	| otherwise = n

div2' :: Int -> Int
div2' n = myIf (even n) (div2' $ n `div` 2) n

myIf :: Bool -> Int -> Int -> Int
myIf True t _ = t
myIf _ _ e = e

notEven :: Int -> Int -> Int
notEven a b
	| even a = b
	| otherwise = a

div2'' :: Int -> Int
div2'' n = notEven n (div2'' $ n `div` 2)

not23 :: Int -> Int -> Int -> Int
not23 a b c
	| even a = b
	| a `mod` 3 == 0 = c
	| otherwise = a

div23 :: Int -> Int
div23 n = not23 n (div23 $ n `div` 2) (div23 $ n `div` 3)
